feat(c/driver_manager): add new function to allow loading by manifest

zeroshade opened the pull request:
Following the discussions on the mailing list and on https://docs.google.com/document/d/1ANeoDI7MCi9xFt_q8z726xKweWS2vgqO8Ut6BSTt5zM/edit?tab=t.0#heading=h.rd6unnsl7uy this PR implements the proposed behavior by adding a new function to the driver manager rather than changing the existing AdbcLoadDriver (for now).

The new function, `AdbcFindLoadDriver` adds a parameter for LoadOptions along with behavior to search directories at multiple levels (environment, user, system) to search for either a manifest or a driver library. This should improve the developer experience and make it easier for users and applications to leverage the driver manager.

zeroshade reviewed (commented):
duplicate comment, oops

zeroshade commented:
@paleolimbot I think i'm gonna need a little help getting the R build to recognize the vendored toml++ header, as far as I can tell looking at `bootstrap-c.R` it should be copying it over correctly from the vendor dir, so I'm not sure what's wrong. Any ideas?

kou commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -138,6 +320,176 @@ struct ManagedLibrary {
+      // if the extension is .toml, attempt to load the manifest
+      // erroring if we fail
+#ifdef _WIN32
+      if (strncasecmp(driver_path.extension().string().c_str(), ".toml", 5) == 0) {
```
Does this match with `.tomlX` too?

zeroshade replied:
added a helper function to manage extension comparison to prevent that

kou commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -138,6 +320,176 @@ struct ManagedLibrary {
+
+#if defined(_WIN32)
+      static const std::string kPlatformLibrarySuffix = ".dll";
+      if (strncasecmp(driver_path.extension().string().c_str(),
+                      kPlatformLibrarySuffix.c_str(),
+                      kPlatformLibrarySuffix.size()) == 0) {
```
`.DLLX` may be matched too.

It may be better that we have a helper function that compares a path extension.

zeroshade commented on c/include/arrow-adbc/adbc_driver_manager.h:
```diff
@@ -31,6 +31,16 @@ extern "C" {
 #ifndef ADBC_DRIVER_MANAGER_H
 #define ADBC_DRIVER_MANAGER_H
 
+enum LoadOptions {
```
Yea, that's why I did that. I can change it to `#define` for consistency, sure

lidavidm commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -15,26 +15,41 @@
-#include <dlfcn.h>
-#endif  // defined(_WIN32)
+ADBC_EXPORT
+std::vector<std::filesystem::path> AdbcParsePath(const std::string_view& path);
```
Can we use `InternalAdbc` instead?

zeroshade replied:
It has to start with `Adbc` because of the symbols.map, so we can't use `InternalAdbc` unless we change the symbols.map 

We could change it to `AdbcInternal` if you like, or just not have the separate unit test for it and not export it, but I prefer to test it

zeroshade replied:
Ah, must have missed that. Why doesn't the other internal function use that too then?

zeroshade replied:
Right, I'm talking about `AdbcDriverManagerDefaultEntrypoint` which is the other function we only expose for unit testing. It's not using `InternalAdbc` that's why I followed the same pattern. It feels like either both should be `InternalAdbc` or neither should

lidavidm replied:
Sure, we can rename it. That probably predates the symbols.map.

zeroshade replied:
Sure, I'll rename both then

lidavidm commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -114,7 +129,175 @@ struct OwnedError {
+  try {
+    config = toml::parse_file(driver_manifest.native());
+  } catch (const toml::parse_error& err) {
+    SetError(error, err.what());
```
Can we have the usual `[DriverManager]` prefix to the error messages?

lidavidm commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -114,7 +129,175 @@ struct OwnedError {
+    config = toml::parse_file(driver_manifest.native());
+  } catch (const toml::parse_error& err) {
+    SetError(error, err.what());
+    return ADBC_STATUS_INVALID_STATE;
```
Also maybe `INVALID_ARGUMENT` is more reasonable?

zeroshade replied:
good point!

lidavidm commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -632,6 +985,36 @@ static const char kDefaultEntrypoint[] = "AdbcDriverInit";
+  constexpr char delimiter = ':';
+#endif  // _WIN32
+
+  bool in_quotes = false;
```
Hmm, is escaping actually a thing? https://unix.stackexchange.com/questions/579081/why-assign-export-ld-library-path suggests it conventionally isn't. (Also it notes that Linux can use `;` too.)

zeroshade replied:
The escaping is needed if there are any spaces in the path. Much more common on Windows than on Linux, but still possible.

I was following the same logic as Go's os package for parsing path delimiters (: for Linux, ; for Windows)

lidavidm replied:
Er, I don't understand. Escaping spaces is the responsibility of the shell, not us

zeroshade replied:
The other reason for needing the quote handling is if you have a path which contains the path separator (: or ;) in the folder itself.

lidavidm replied:
Right, but apparently conventionally this isn't handled, so why are we inventing a convention? (Also how do you escape quotes? And is there any precedent for doing this vs a backslash?)

lidavidm replied:
Go only does that on Windows, though

lidavidm replied:
I can't find a good reference but it seems this is implicit/expected behavior on Windows, but isn't on Linux. If we're going to do this we should at least point to the reference being used for the behavior

lidavidm replied:
I assume the goal was to follow platform conventions in the first place. That includes not bleeding over behavior from one platform to another

zeroshade replied:
Sure, I'll update this then and add a comment linking to the reference. 

There's surprisingly few examples of prior art here that I can find

zeroshade replied:
I also expect others may end up being more opinionated on this than us.... Lol

zeroshade replied:
In other discussions we've come to the conclusion to avoid exposing a version that takes the list of strings. I'll update this to work more similar to https://cs.opensource.google/go/go/+/refs/tags/go1.24.3:src/path/filepath/path_windows.go (only do the quotes/escaping check on windows, etc.)

lidavidm commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -1658,8 +2041,76 @@ const char* AdbcStatusCodeMessage(AdbcStatusCode code) {
+
 AdbcStatusCode AdbcLoadDriver(const char* driver_name, const char* entrypoint,
                               int version, void* raw_driver, struct AdbcError* error) {
+  // should this just be AdbcFindLoadDriver(...., kAdbcDefaultLoadOptions, ...)?
```
yeah, probably

pitrou commented on c/driver_manager/adbc_driver_manager_test.cc:
```diff
@@ -335,4 +338,31 @@ TEST(AdbcDriverManagerInternal, AdbcDriverManagerDefaultEntrypoint) {
+
+  std::ostringstream joined;
+  std::copy(paths.begin(), paths.end(),
+            std::ostream_iterator<std::string>(joined, delimiter));
```
Apparently, `ostream_iterator` will also append a delimiter at the end (see example):
https://en.cppreference.com/w/cpp/iterator/ostream_iterator.html

zeroshade replied:
yea, which is fine for the test because it ensures we also remove empty paths from the returned list

pitrou commented on go/adbc/drivermgr/adbc_driver_manager.cc:
```diff
@@ -114,7 +128,175 @@ struct OwnedError {
   }
 };
 
+std::filesystem::path UserConfigDir() {
```
This is copy-pasted from the C ADBC codebase??

zeroshade replied:
I don't think so? I haven't seen this anywhere else, is there a spot that already does this?

zeroshade replied:
ah, yea. I didn't realize this comment was on the Go copy.

Yea, we sync the code for the driver manager to the Go package so that it builds and statically links the driver manager into the Go driver_manager package. For `go get` you can only reference files that are in the actual Go module code base, so we can't just point it at the C code directly if we want it to be easily usable in other go projects.

zeroshade commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -114,7 +128,175 @@ struct OwnedError {
+std::vector<std::filesystem::path> GetSearchPaths(const uint16_t levels) {
+  std::vector<std::filesystem::path> paths;
+  if (levels & static_cast<uint16_t>(kAdbcSearchEnvironment)) {
+    // Check the ADBC_DRIVER_PATH environment variable
+    const char* env_path = std::getenv("ADBC_CONFIG_PATH");
```
Fixed

zeroshade commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -114,7 +128,175 @@ struct OwnedError {
+  info.source = config["source"].value_or(""s);
+  auto lib = config.at_path("Driver.shared").value<std::string>();
+  if (!lib) {
+    SetError(error, "Driver path not found in manifest");
```
added

felipecrv commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -114,7 +129,175 @@ struct OwnedError {
+std::filesystem::path UserConfigDir() {
+  std::filesystem::path config_dir;
+#if defined(_WIN32)
+  auto dir = std::getenv("AppData");
```
> ...create a Windows user with a é (e.g.)

Everyone goes through this ritual with Windows programming, huh? :)

zeroshade replied:
@paleolimbot I don't think we need to use the Windows API that you're pointing at there, as the only available options for that are %LOCALAPPDATA%\Desktop, %LOCALAPPDATA%\Documents, %LOCALAPPDATA%\Favorites and %LOCALAPPDATA%\ProgramData so it looks like you can't just get the `%AppData%` folder with that API

felipecrv commented on go/adbc/drivermgr/adbc_driver_manager.cc:
```diff
@@ -114,7 +125,187 @@ struct OwnedError {
+  HKEY key() const { return key_; }
+  bool is_open() const { return is_open_; }
+
+  std::string GetString(const std::string& name, const std::string& default_value = "") {
```
Drop the default value to make it explicit at the callsite that `""` will return if not found and pass by value, since the function can consume the string when returning.

```suggestion
  std::string GetString(const std::string& name, std::string default_value) {
```

zeroshade replied:
I can understand changing it to not have a default value, but is the intent of passing by value there just so we can potentially elide the copy for the return via move semantics?

felipecrv commented on go/adbc/drivermgr/adbc_driver_manager.cc:
```diff
@@ -114,7 +125,187 @@ struct OwnedError {
+  RegistryKey(HKEY root, const std::string_view subkey) noexcept
+      : root_(root), key_(nullptr), is_open_(false) {
+    auto result = RegOpenKeyExA(root_, subkey.data(), 0, KEY_READ, &key_);
+    is_open_ = (result == ERROR_SUCCESS);
```
Better to retain the error as a member variable and define a `bool is_open() const` that performs this comparison and a getter for the result can improve the error handling.

felipecrv commented on go/adbc/drivermgr/adbc_driver_manager.cc:
```diff
@@ -114,7 +125,187 @@ struct OwnedError {
+  info.lib_path = dkey.GetString("driver");
+  if (info.lib_path.empty()) {
+    SetError(error,
+             "[DriverManager] Driver path not found in registry for "s + driver_name);
```
Just in case `driver_name` is empty.

```suggestion
             "[DriverManager] Driver path not found in registry for \""s + driver_name + "\"");
```

felipecrv commented on go/adbc/drivermgr/adbc_driver_manager.cc:
```diff
@@ -1658,9 +2047,10 @@ const char* AdbcStatusCodeMessage(AdbcStatusCode code) {
-AdbcStatusCode AdbcLoadDriver(const char* driver_name, const char* entrypoint,
-                              int version, void* raw_driver, struct AdbcError* error) {
-  AdbcDriverInitFunc init_func;
+AdbcStatusCode AdbcFindLoadDriver(const char* driver_name, const char* entrypoint,
```
Very nice!

felipecrv commented on go/adbc/drivermgr/arrow-adbc/adbc_driver_manager.h:
```diff
@@ -51,6 +63,52 @@ ADBC_EXPORT
+ADBC_EXPORT
+AdbcStatusCode AdbcFindLoadDriver(const char* driver_name, const char* entrypoint,
+                                  const int version, const AdbcLoadOption load_options,
+                                  void* driver, struct AdbcError* error);
```
Thanks for exposing `load_options`!

felipecrv commented on c/include/arrow-adbc/adbc_driver_manager.h:
```diff
@@ -31,6 +32,17 @@ extern "C" {
 #ifndef ADBC_DRIVER_MANAGER_H
 #define ADBC_DRIVER_MANAGER_H
 
+typedef uint16_t AdbcLoadOption;
```
Given how "option" is overloaded in ADBC already, this could be called `AdbcLoadFlags`.

`uint32_t` is also a better default type for this.

zeroshade commented on c/vendor/toml++/toml.hpp:
the release build of the driver manager with these changes and updates is only 320K, so I don't think this will necessarily result in people opting out of the driver manager. Unless I'm missing something?

zeroshade replied:
I don't know much about vcpkg configs but I can probably set up some option so that it can use the vcpkg version instead of the vendored one? Right?

zeroshade commented on c/include/arrow-adbc/adbc_driver_manager.h:
```diff
@@ -51,6 +63,52 @@ ADBC_EXPORT
+ADBC_EXPORT
+AdbcStatusCode AdbcFindLoadDriver(const char* driver_name, const char* entrypoint,
+                                  const int version, const AdbcLoadOption load_options,
+                                  void* driver, struct AdbcError* error);
```
We make no thread safety guarantees during loading a driver. I don't think we can anyways because of dlopen using LD_LIBRARY_PATH. 

if you're programmatically changing the env vars in one thread while loading a driver in another thread, that's on you to ensure safety lol

At least that's my opinion.

felipecrv replied:
I think adding a list of paths here overcomplicates things.

Any user of the driver manager that wants to use their own location should implement their own mechanism and build the absolute path. And path resolution issue becomes their responsibility.

The behavior of the adbc driver manager should be easy to explain and we should incentivize applications to follow the conventional paths it looks for instead of making it infinitely configurable with a path list.

zeroshade replied:
If you're passing an absolute path, then we don't check `ADBC_CONFIG_PATH` either, so using a full path eliminates the cases where we'd rely on the env var, so it solves both cases.

> Providing the search path would ensure a means by which a caller could do this safely (otherwise they would have to opt out of the manifests entirely or write their own parsing logic?)

Is there really a use case where a user would even want to be programmatically changing the ADBC_CONFIG_PATH in a thread while simultaneously loading drivers? Couldn't they just set the search path into the env var first before doing anything?

lidavidm replied:
Yeah, I think an application that is very concerned about this isn't going to want the rest of the mechanics in the driver manager anyways and can manage its own paths

zeroshade commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -114,7 +128,175 @@ struct OwnedError {
+  info.source = dkey.GetString("source");
+  info.lib_path = dkey.GetString("driver");
+  if (info.lib_path.empty()) {
+    SetError(error, "Driver path not found in registry");
```
added

zeroshade commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -114,7 +128,175 @@ struct OwnedError {
+ public:
+  RegistryKey(HKEY root, const std::string_view subkey) noexcept
+      : root_(root), key_(nullptr), is_open_(false) {
+    auto result = RegOpenKeyExA(root_, subkey.data(), 0, KEY_READ, &key_);
```
Updated to use the `xxxW()` methods

lidavidm commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -114,7 +125,187 @@ struct OwnedError {
+  }
+#endif
+
+  return config_dir;
```
> I think you will want this handled for the "people running R/Python in Windows on a Surface" use case, who probably have no idea what architecture they're running

Is the concern basically, they may have installed the amd64 driver while on an ARM processor (or vice versa) without realizing?

lidavidm replied:
I suppose it doesn't hurt to either note the platform/architecture in the manifest or in the filepath somehow.

zeroshade replied:
Oh hell, do we really need it to be triples instead of just "arm64" / "amd64" etc? 

I would imagine that you wouldn't distribute a single manifest with multiple platform builds? Or am I wrong and we need it to be full triples?

zeroshade replied:
well that's just awful....

okay how about the following:

`(windows|osx|linux|unknown)-(amd64|arm64|x86|arm|s390x|unknown)[-musl|-mingw]`

what do you think?

zeroshade commented:
@pitrou @paleolimbot @lidavidm Any further comments here?

lidavidm reviewed (commented):
Yeah, I think we just need to see tests now.

lidavidm commented on c/driver_manager/current_arch.h:
```diff
@@ -0,0 +1,84 @@
+
+#include <string>
+
+const std::string& current_arch() {
```
well once we have c++20 we can just format everything...and if we follow Arrow's lead that should be soon. oh well then

lidavidm commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -114,7 +128,228 @@ struct OwnedError {
+#ifdef _WIN32
+using char_type = wchar_t;
+
+std::string utf8_encode(const std::wstring& wstr) {
```
This isn't resolved? We should try to follow the existing convention

zeroshade commented:
@paleolimbot looks like @daipom already identified the issue in https://github.com/apache/arrow-adbc/issues/2993 so I think it can be ignored on this PR until that gets fixed. My change here definitely isn't causing it :smile:

zeroshade commented:
@paleolimbot @lidavidm @kou @pitrou any further comments here?

kou commented on c/cmake_modules/AdbcDefines.cmake:
```diff
@@ -90,6 +89,8 @@ if(MSVC)
   add_compile_options(/wd4868)
   add_compile_options(/wd4710)
   add_compile_options(/wd4711)
+  # Don't warn about enforcing left-to-right evaluation order for operator[]
+  add_compile_options(/wd4866)
   # Don't warn about padding added after members
   add_compile_options(/wd4820)
```
We may want to sort by warning ID:

```suggestion
  # Don't warn about padding added after members
  add_compile_options(/wd4820)
  # Don't warn about enforcing left-to-right evaluation order for operator[]
  add_compile_options(/wd4866)
```

kou commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -15,26 +15,42 @@
 
-#include "arrow-adbc/adbc_driver_manager.h"
+#if defined(_WIN32)
+#pragma comment(lib, "advapi32.lib")  // for registry stuff
```
Can we move this to `CMakeiLists.txt`?

kou commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -15,26 +15,42 @@
+#include <string.h>  // _wcsnicmp
+#include <strsafe.h>
+
+#include <codecvt>
```
This is deprecated in C++17...

lidavidm replied:
Since we only need the conversions on Windows, maybe we can use Win32 instead? https://learn.microsoft.com/en-us/previous-versions/bb202786(v=msdn.10)?redirectedfrom=MSDN

kou commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -15,26 +15,42 @@
 #include <cctype>
 #include <cerrno>
 #include <cstring>
+#include <filesystem>  // NOLINT [build/c++17]
```
Can we remove this comment by #2989?

zeroshade replied:
good point!

kou commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -114,7 +130,186 @@ struct OwnedError {
+std::string Utf8Encode(const std::wstring& wstr) {
+  if (wstr.empty()) return std::string();
+  int size_needed = WideCharToMultiByte(
+      CP_UTF8, 0, &wstr[0], static_cast<int>(wstr.size()), NULL, 0, NULL, NULL);
```
```suggestion
      CP_UTF8, 0, wstr.data(), static_cast<int>(wstr.size()), NULL, 0, NULL, NULL);
```

kou commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -114,7 +130,186 @@ struct OwnedError {
+  int size_needed = WideCharToMultiByte(
+      CP_UTF8, 0, &wstr[0], static_cast<int>(wstr.size()), NULL, 0, NULL, NULL);
+  std::string str_to(size_needed, 0);
+  WideCharToMultiByte(CP_UTF8, 0, &wstr[0], static_cast<int>(wstr.size()), &str_to[0],
```
```suggestion
  WideCharToMultiByte(CP_UTF8, 0, wstr.data(), static_cast<int>(wstr.size()), str_to.data(),
```

kou commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -114,7 +130,186 @@ struct OwnedError {
+std::wstring Utf8Decode(const std::string& str) {
+  if (str.empty()) return std::wstring();
+  int size_needed =
+      MultiByteToWideChar(CP_UTF8, 0, &str[0], static_cast<int>(str.size()), NULL, 0);
```
```suggestion
      MultiByteToWideChar(CP_UTF8, 0, str.data(), static_cast<int>(str.size()), NULL, 0);
```

kou commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -114,7 +130,186 @@ struct OwnedError {
+  int size_needed =
+      MultiByteToWideChar(CP_UTF8, 0, &str[0], static_cast<int>(str.size()), NULL, 0);
+  std::wstring wstr_to(size_needed, 0);
+  MultiByteToWideChar(CP_UTF8, 0, &str[0], static_cast<int>(str.size()), &wstr_to[0],
```
```suggestion
  MultiByteToWideChar(CP_UTF8, 0, str.data(), static_cast<int>(str.size()), wstr_to.data(),
```

kou commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -114,7 +130,186 @@ struct OwnedError {
+
+    DWORD type = REG_SZ;
+    DWORD size = 0;
+    auto result = RegQueryValueExW(key_, name.c_str(), nullptr, &type, nullptr, &size);
```
We can use `data()` with C++11 or later because `data()` with C++11 or later returns `NULL` terminated C string:

```suggestion
    auto result = RegQueryValueExW(key_, name.data(), nullptr, &type, nullptr, &size);
```

kou commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -114,7 +130,186 @@ struct OwnedError {
+    if (type != REG_SZ) return default_value;
+
+    std::wstring value(size, '\0');
+    result = RegQueryValueExW(key_, name.c_str(), nullptr, &type,
```
```suggestion
    result = RegQueryValueExW(key_, name.data(), nullptr, &type,
```

kou commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -114,7 +130,186 @@ struct OwnedError {
+  auto wext = Utf8Decode(ext);
+  auto path_ext = path.extension().native();
+  return path_ext.size() == wext.size() &&
+         _wcsnicmp(path_ext.c_str(), wext.c_str(), wext.size()) == 0;
```
```suggestion
         _wcsnicmp(path_ext.data(), wext.data(), wext.size()) == 0;
```

kou commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -138,22 +333,189 @@ struct ManagedLibrary {
+
+        auto status = LoadDriverManifest(driver_path, info, error);
+        if (status == ADBC_STATUS_OK) {
+          return Load(info.lib_path.c_str(), error);
```
```suggestion
          return Load(info.lib_path.data(), error);
```

kou replied:
(We have more `.c_str()`.)

zeroshade commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -114,7 +130,186 @@ struct OwnedError {
+  WideCharToMultiByte(CP_UTF8, 0, &wstr[0], static_cast<int>(wstr.size()), &str_to[0],
+                      size_needed, NULL, NULL);
+  return str_to;
+}
```
latest change added a test with a non-ascii file path being passed through the environment var

zeroshade commented:
Okay, I think I addressed all of the comments from everyone. I'll update the docs in a follow-up PR. 

Please take a look and lemme know if I forgot anything. Thanks!!

lidavidm commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -15,26 +15,45 @@
 
-#include "arrow-adbc/adbc_driver_manager.h"
+#if defined(_WIN32)
+#include <windows.h>  // Must come first
```
should we define _NOMINMAX too just to be safe?

zeroshade replied:
When I google `_NOMINMAX` it looks like the pre-processor macro doesn't have the leading underscore in the results. Is it supposed to have the underscore or not?

Also, I probably wouldn't want to add the define unless we actually do encounter the issue and need it personally. Thoughts?

lidavidm commented on c/driver_manager/adbc_driver_manager.cc:
```diff
@@ -1716,6 +2176,12 @@ AdbcStatusCode AdbcLoadDriver(const char* driver_name, const char* entrypoint,
   return status;
 }
 
+AdbcStatusCode AdbcLoadDriver(const char* driver_name, const char* entrypoint,
+                              int version, void* raw_driver, struct AdbcError* error) {
+  return AdbcFindLoadDriver(driver_name, entrypoint, version, ADBC_LOAD_FLAG_DEFAULT,
+                            raw_driver, error);
+}
```
Maybe this should default to no options enabled?

zeroshade replied:
I would argue we should default to all options enabled so it searches everywhere for the manifests and applications that use the driver would have to explicitly disable search locations (opt-out, not opt-in). This way it ensures that the default behavior is that the manifest files work and are easily found.

lidavidm reviewed (approved):
Just some nits

zeroshade commented:
Thanks everyone for all the comments and work here! Glad to get this in. I'll follow this up with changes for the docs and updates to the bindings
